### DISCUSSION TITLE
How difficult would it be to implement Secure Remote Password protocol (SRP)?

### DISCUSSION THREAD

**Discussion by jesse-savary** [keycloak/keycloak#12536] [Q&A]
Hello,

I'm currently evaluating Keycloak as a replacement for Cognito in our stack and was wondering how difficult it would be to add SRP. I'm not familiar with Keycloak's internals and so I'm not sure if I could build this as some kind of plugin or if I would have to modify the core directly.

Any input is appreciated, more details on SRP here: https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol

**Comment by korneliuscode** [comment]
I’m actually looking for the same answer. Did you find anything on this @jesse-savary?

**Comment by ddurham2** [comment]
I think your question may need some more detail.  In what context are you asking for SRP?  e.g. Are you asking for it not to exchange the creds directly between client and server when user must enter the password, or something else?  Or else, describe in what context Cognito uses SRP (supposing that's what you're after).

**Reply by korneliuscode** [reply]
I’m not looking to replace anything but to introduce keycloak into our multi tenancy SaaS application. 
Within that - as you say - I would love to use SRP to not send and/or store any password equivalent data for the user during login/register etc.

**Reply by sschu** [reply]
Keycloak stores user passwords in its database hashed/salted and generates tokens for securely accessing resources. What is your use case to store passwords anywhere else?

**Reply by korneliuscode** [reply]
[@sschu](https://github.com/sschu) the use case for SRP is to not even store hased/salted passwords in the database! SRP is a standard for not storing any password-equivalent data on the server. If someone would obtain the database there would be no way to (beyond hased/saled passwords) to leverage that as an attach vector. Also it helps again man-in-the-middle attacks. Read more [here](https://en.wikipedia.org/wiki/Secure_Remote_Password_protocol).

**Reply by sschu** [reply]
I still don't get what you are trying to achieve. Keycloak is a web application used via web browsers. To authenticate users have to enter passwords into forms and submit it. I would be surprised if there was a HTTP binding and web browsers supporting SRP to prove credential ownership and there is no mention of this in the linked Wikipedia article but I am clearly not an expert here.

**Reply by ecasilla** [reply]
https://www.youtube.com/watch?v=RWksEY-Bf9I This video also recommends to use keycloak even though keycloak doesn't implement SRP 😅

**Reply by stianst** [reply]
SRP is just a challenge-response thing that means the server can verify the users password, without having the actual password (or hash representation of the password). As such it's kinda nifty idea to use that in Keycloak.

Implementation wise it's probably something that can be POC relatively easily with a custom authenticator and custom credential type in Keycloak. You'd of course need some JavaScript added to the client-side, which the authenticator can do.

I tend to disagree with the man in the middle attack being all that valid here. End-to-end TLS takes care of that just fine. If anyone are able to intercept the users passwords with a mitm they can just as well sniff tokens instead, so proper TLS is a must regardless.

Rather the real thing this would add additional protection against is ability to obtain users passwords if the Keycloak server or its database is compromised. It does rely on users reusing the same password cross multiple services to be useful though. However, there are other things that protect against that. Not re-using passwords cross different services is one, two-factor authentication another, WebAuthN is a great one (as it basically is similar to SRP with it being a challenge-response thing, but compared to SRP significantly stronger as it doesn't rely on potentially short/bad passwords). 

So, all in all I'm not all that convinced if this stuff is needed.

**Reply by korneliuscode** [reply]
@stianst while I agree with what you say, your suggestion rely mostly on the users doing the right thing. I learned this is the hardest thing and it does not work if you have less tech savvy users. I believe the platform should to everything to keep it simple and secure for the users. Which is why I would love to see SRP or an equivalent in keycloak.

**Reply by naturallight** [reply]
@stianst SRP (Secure Remote Password protocol) is designed for a different threat model. In real production environments, end-to-end TLS usually terminates at the API Gateway (e.g., NGINX). From there, the password might pass through multiple intermediary nodes before reaching the Keycloak service, which increases the attack surface. If any of these nodes are compromised, there is a risk of password leakage.

SRP addresses this threat model effectively. Even if intermediary nodes are compromised, an attacker still cannot retrieve the actual password.

**Reply by rustonaut** [reply]
While the different threat model exists and some form of  Asymmetric/Argumented  Password-Authenticated Key Exchange (aPAKE) protocols like SRP can help with it, I personally still wouldn't add support for SRP.

(Disclaimer: I'm at best a security aware software engineer, not a security researcher or cryptographer)

A non exhaustive lists of issues with SRP:

- even through SRP is kinda old and seems simple the degree of proven security guarantees compared to other aPAKEs is surprisingly limited 
- there are better alternatives e.g. OPAQUE
- SRP might not fix your thread model even if it might look like it initially
  - e.g. the `H(salt, password)` send during registration is basically like a "high entropy password" anyone who gets it can impersonate the  client and snoop on conversations encrypted with the SRP derived key. This means if you thread model was "what if someone can read the communication between web gateway and server" it isn't really fully fixed by SRP. Sure they don't get the password only the hash, but you also never get the password only auth based on the hash => so the hash **is** the password. 
  - also I might be missing something but in difference to more modern aPAKEs liek OPAQUE, SRP seems to be missing a non-ephemeral server key, I'm not fully sure about the exact consequences but it should make impersonating the server way easier (in context of having "seen" the registration procedure and/or MITM the registration procedure)
  - if you thread model full E2EE between client and backend services accessed by request first passing through other services (e.g. web gateway),  things get **supper complicated**, SRP (or any aPAKE) will not safe you here at all
  
all in all this means even if Keycloak would implement an aPAKE for password login, it probably would not be SRP

but there is a much bigger problem with Keycloak and aPAKE in general

- the idea of aPAKE isn't just login, but to produce a shared secret you then use for encrypting a communication channel 
  - as consequence you normally want to run the aPAKE between the client and server which do actually communicate for the non-login interaction i.e. in OIDC terms the relying party (client) and resource server (i.e. not Keycloak)
  - i.e. aPAKE would replace the `Authroization: Bearer <access_token>` part in resource servers but that means it really doesn't work that well any of: SSO, OIDC, OAuth2
  - but "SSO, OIDC, OAuth2" is kinda the main point of Keycloak :sweat_smile: 
  
and if using a aPAKE only for login but not key exhange then the go to solution this is competing with for "having some MITM/snooping protections on login" seems to have become passkeys, even if I don't like the direction they developed into  :disappointed:  

now to be clear the idea of not even sending password so that no one can snoop on it still is nice

**Reply by ocontant** [reply]
What would be your point of view in regards to a Vault use case, where there is a requirement for Zero Knowledge where the provider must not know the password or the secret key to decrypt the vault?
